**Title**:

 Analysis of Naive Bayes and Support Vector Machine for Text Classification.

 **Abstract:**



 This lab report explores the performance of Naive Bayes (NB) and Support Vector Machine (SVM) algorithms in the context of Bitcoin Price Prediction classification. Using a dataset of Bitcoin Price Prediction, we implemented and evaluated both algorithms, considering their accuracy and computational efficiency. The results indicate distinct strengths and weaknesses for each method.



In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import StandardScaler
from keras.models import Sequential
from keras.layers import Dense
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import MinMaxScaler

data = pd.read_csv('/content/BTC-USD.csv')


data['Date'] = pd.to_datetime(data['Date'])

data = data.sort_values('Date')

missing_values = data.isnull().sum()
print("Missing Values:")
print(missing_values)


data.fillna(method='ffill', inplace=True)


scaler = MinMaxScaler()
numerical_columns = ['Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume']
data[numerical_columns] = scaler.fit_transform(data[numerical_columns])

data['Target'] = data['Close'].shift(-1)

data = data.dropna()


data.to_csv('preprocessed_dataset.csv', index=False)


Missing Values:
Date         0
Open         0
High         0
Low          0
Close        0
Adj Close    0
Volume       0
dtype: int64


**Naive Bayes implementation:**



The Naive Bayes algorithm was implemented using the scikit-learn library in Python. The Multinomial Naive Bayes variant was chosen due to its suitability for text data. Hyperparameter tuning was performed to optimize the model.



In [ ]:


# Load your dataset (replace 'your_dataset.csv' with the actual file path)
data = pd.read_csv('/content/preprocessed_dataset.csv')

# Encoding the "Target" variable as binary (0 for price decrease, 1 for price increase)
data['Target'] = (data['Target'] > data['Close']).astype(int)

X = data[['Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume']]
y = data['Target']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

nb_classifier = GaussianNB()
nb_classifier.fit(X_train, y_train)

y_pred = nb_classifier.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f"NB Accuracy: {accuracy * 100:.2f}%")



NB Accuracy: 53.78%


**Support Vector Machine Implementation:**


The Support Vector Machine was implemented using the linear kernel for simplicity. Grid search was employed to fine-tune the C parameter, balancing between margin maximization and classification error.



In [ ]:
# Create and train an SVM classifier with a linear kernel
svm_classifier = SVC(kernel='linear', C=1)
svm_classifier.fit(X_train, y_train)

# Make predictions on the test set
y_pred = svm_classifier.predict(X_test)

# Evaluate the model's accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"SVM Accuracy: {accuracy * 100:.2f}%")


SVM Accuracy: 56.35%


**Experimental Results:**


The results revealed that Naive Bayes achieved an accuracy of 53.78%, while Support Vector Machine achieved 56.35%. Although SVM outperformed NB in this experiment, it's essential to consider other factors such as interpretability and computational efficiency.

**Conclusion:**

In conclusion, this experiment provided insights into the performance of Naive Bayes and Support Vector Machine in Bitcoin Price Prediction classification. The choice between these algorithms depends on the specific requirements of the task at hand, considering factors such as accuracy, interpretability, and computational efficiency.